core/src/main/java/com/digitalpebble/stormcrawler/bolt/JSoupParserBolt.java

```diff
@@ -489,7 +486,7 @@ public String guessMimeType(String URL, String httpCT, byte[] content) {
             // the URL is valid
             LOG.error("MalformedURLException on {}", url);
             eventCounter.scope("error_invalid_source_url").incrBy(1);
-            return new LinkedList<Outlink>();
+            return new LinkedList<>();
```

Comment:
I left out the explicit type argument of the LinkedList because the return type is already defined as `List<Outlink>`, therefore the compiler infers the parametertype from the signature via the diamond operator <>. It works basically the same as declaring a list like that: `List<String> myList = new ArrayList<>()`.

The other reason is, that the explicit type argument of a generic is resolved during compile time and then erased. Therefore the type argument of a generic doesn't add anything to the class itself during runtime. 
That's the same reason for the fact that you could return a raw type like `new LinkedList()` and it would still compile and work the same as returning `LinkedList<>` or `LinkedList<Outlink>` as long as your return type is `List<Outlink>`. Basically every generic class that you use, loses most\* of it's information about its generic types during compile time and replaces them in the bytecode (conceptually, the real results of the compiler might differ due to optimisations and other whatnots) like this:

```java
List<String> = new ArrayList<>();
list.add("A value!");
String s = list.get(0);

List list = new ArrayList();
list.add("A value!");
String s = (String) list.get(0);
```

And the explicit declaration of generics in a signature like `Map<String, Object>` has multiple advantages over raw types:
- The compiler creates and optimizes the conversion code of the values for you and you don't have to do that by yourself. There is also the possibility, that the compiler doesn't introduce any type checking during runtime at all.
- The interface is clearer for the reader because the reader sees what's expected from the generic objects like a Map or List without knowing every inch of the code. Like in the following example, where it is easier to implement something when the interface is declared like in framework B:
```java

// Framework A
public abstract class Event {
    public abstract String eventMessage();
}

@FunctionalInterface
public interface EventConsumer {
    public void consume(List events);
}

public interface EventSource {
    public void register(EventConsumer event);
}

// Usage in code
public class NetworkListener implements EventSource {
    // This is the only Event sent by this class.
    public static class NetworkEvent extends Event {
      //...
    }

    public void register(EventConsumer event) {
      //...
    }
}

// Framework B
public abstract class Event {
    abstract String eventMessage();
}

@FunctionalInterface
public interface EventConsumer<T extends Event> {
    void consume(List<T> events);
}

public interface EventSourcee<T extends Event> {
    public void register(EventConsumer<T> event);
}

// Usage in code
public class NetworkListener implements EventSource<NetworkListener.NetworkEvent> {
     // This is the only Event sent by this class.
    public static class NetworkEvent extends Event {
        //...
    }

    public void register(EventConsumer<NetworkEvent> event) {
        //...
    }
}

```
- That you can declare raw types is a legacy of old java code prior to jdk 5.0 when there was no "generic". Therefore when you use generics you get essentially pre generic behaviour and that is sometimes really strange and not wanted or even dangerous.

Sources: 
- https://www.baeldung.com/java-type-erasure
- https://docs.oracle.com/javase/tutorial/java/generics/rawTypes.html
- https://docs.oracle.com/javase/tutorial/java/generics/erasure.html

\* I don't want to say "all" because I'm not deep enough in the "java type erasure during compilation"-process to know if java really erases all informations about the generic type from the class instance.

Reply:
thanks for the detailed explanation. Makes sense